barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/AvmMini_trace.cpp

```diff
@@ -522,6 +535,74 @@ std::vector<FF> AvmMiniTraceBuilder::returnOP(uint32_t retOffset, uint32_t retSi
     return returnMem;
 }
 
+// TODO: doc
+void AvmMiniTraceBuilder::halt()
+{
+    auto clk = mainTrace.size();
+
+    mainTrace.push_back(Row{
+        .avmMini_clk = clk,
+        .avmMini_pc = FF(pc),
+        .avmMini_internal_return_ptr = FF(internal_return_ptr),
+        .avmMini_sel_halt = FF(1),
+    });
+}
+
+void AvmMiniTraceBuilder::internal_call(uint32_t jmpDest)
+{
+    auto clk = mainTrace.size();
+
+    uint32_t stored_pc = pc + 1;
+    internal_call_stack.push(stored_pc);
+
+    // We want to write the current pc to the memory location pointed by
+    storeAInMemTrace(internal_return_ptr, FF(stored_pc));
+
+    mainTrace.push_back(Row{
+        .avmMini_clk = clk,
+        .avmMini_pc = FF(pc),
+        .avmMini_internal_return_ptr = FF(internal_return_ptr),
+        .avmMini_sel_internal_call = FF(1),
+        .avmMini_ia = stored_pc,
+        .avmMini_mem_op_a = FF(1),
+        .avmMini_rwa = FF(1),
+        .avmMini_mem_idx_a = FF(internal_return_ptr),
+    });
+
+    // We want the next row to be the one pointed by jmpDest
+    ffMemory.at(internal_return_ptr) = stored_pc;
```

Comment:
@Maddiaa0 We should move this line just after 
` storeAInMemTrace(internal_return_ptr, FF(stored_pc));`